### PR TITLE
[region-isolation] Only make partial_apply actor derived if we capture an isolated parameter... not just any actor.

### PR DESCRIPTION
**Pull request by gottesmm**
Before I couldn't do this since, @sil_isolated was not represented on partial_applies. In the first commit in this commit series, I added support to the frontend so that we now emit partial_applies with sil_isolated parameters.

In the second commit, I use this information to restrict when the partial_apply is considered to be "actor derived" to if the partial_apply parameter is @sil_isolated.

rdar://123881277

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SIL/IR/SILFunctionType.cpp [hunk 1]
```diff
@@ -1917,9 +1917,11 @@ lowerCaptureContextParameters(TypeConverter &TC, SILDeclRef function,
   // signature from the AST for that.
   auto origGenericSig = function.getAnyFunctionRef()->getGenericSignature();
   auto loweredCaptures = TC.getLoweredLocalCaptures(function);
+  auto *isolatedParam = loweredCaptures.getIsolatedParamCapture();
 
   for (auto capture : loweredCaptures.getCaptures()) {
     if (capture.isDynamicSelfMetadata()) {
+      assert(!isolatedParam);
```
Comment (rjmccall) [inline lib/SIL/IR/SILFunctionType.cpp:1924]: What's this assertion for?  Is there a reason we can't have a dynamic self metadata capture *and* an isolated parameter or opaque value?  You don't seem to be using it anyway.

**Review by rjmccall — COMMENTED** [review]
Other than the two questionable assertions, this looks good.

File: lib/SIL/IR/SILFunctionType.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SIL/IR/SILFunctionType.cpp:1924]: I just was being paranoid. I can remove it.

**Comment by gottesmm** [comment]
Problem was that in CaptureInfo we were determining if a value was isLocalCapture by calling CapturedValue::getDecl(). This doesn't work for dynamic self metadata which is never an isolation parameter (it is a metatype) and doesn't have a decl. I am preparing a small addition to this PR with a patch that fixes that issue by adding a new isLocalCapture to CapturedValue that does the correct thing.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test macOS platform

**Pull request merged**